File: apps/scanner/lib/core/checkin.dart
import 'token.dart';

/// The check-in state machine — Dart port of
/// packages/checkin-core/src/checkin.ts.
///
/// Deliberately pure: no database, no clock, no network. Everything it
/// needs arrives in the context — which is what lets identical logic run
/// on a phone with no signal and on the server when the queue syncs.
///
/// The ported test suite is the specification. Behaviour must match the
/// TypeScript original exactly; if they ever disagree, both are wrong.

enum Rsvp { pending, attending, partial, declined }

enum Outcome {
  needsCount('needs_count'),
  admitted('admitted'),
  partial('partial'),
  overflowAdmitted('overflow_admitted'),
  manual('manual'),
  allowanceExhausted('allowance_exhausted'),
  invalid('invalid'),
  wrongEvent('wrong_event'),
  wrongLeg('wrong_leg'),
  revoked('revoked'),
  rsvpBlocked('rsvp_blocked'),
  rsvpDeclined('rsvp_declined'),
  overflowBlocked('overflow_blocked'),
  notFound('not_found');

  /// The string written to check_in_events.result on the server.
  final String wire;
  const Outcome(this.wire);
}

enum Tone { admit, hold, deny, ask }

/// One household's row in the scanner's local copy of the guest list.
class LocalInvitation {
  final String passId;
  final String invitationId;
  final String eventId;
  final String legId;
  final String displayName;
  final String? category;
  final String? tableName;
  final int allowance;

  /// Derived locally: SUM(admitted_count) for this pass at this leg.
  final int admitted;
  final Rsvp rsvp;
  final bool revoked;

  const LocalInvitation({
    required this.passId,
    required this.invitationId,
    required this.eventId,
    required this.legId,
    required this.displayName,
    this.category,
    this.tableName,
    required this.allowance,
    required this.admitted,
    required this.rsvp,
    required this.revoked,
  });
}

class Policy {
  final bool allowOverflow;
  final bool requireRsvp;
  const Policy({required this.allowOverflow, required this.requireRsvp});
}

class Context {
  final String currentEventId;
  final String currentLegId;
  final Policy policy;

  /// Keys for every event this usher works, not only the current one.
  final List<EventKey> keys;

  /// Local lookup by pass id. Null means not in this leg's list.
  final LocalInvitation? Function(String passId) find;

  /// Usher permission. Off by default.
  final bool canOverrideRsvp;

  const Context({
    required this.currentEventId,
    required this.currentLegId,
    required this.policy,
    required this.keys,
    required this.find,
    this.canOverrideRsvp = false,
  });
}

sealed class ScanInput {
  final int? requestedCount;
  const ScanInput(this.requestedCount);
}

class ScanRaw extends ScanInput {
  final String raw;
  const ScanRaw(this.raw, {int? requestedCount}) : super(requestedCount);
}

/// Name search, then check in by hand. Skips all token checks.
class ManualInput extends ScanInput {
  final String passId;
  const ManualInput(this.passId, {int? requestedCount}) : super(requestedCount);
}

class Decision {
  final Outcome outcome;
  final Tone tone;

  /// Counts against allowance. Zero for every refusal.
  final int admittedCount;
  final String headline;
  final String? detail;

  /// Household, when we could identify one.
  final LocalInvitation? invitation;

  /// Remaining after this decision is applied.
  final int? remaining;

  /// Present on needs_count — what the picker should offer.
  final List<int>? choices;

  /// Whether to write a row. Refusals are logged; the count prompt is not.
  final bool log;

  /// Milliseconds before returning to the camera. Null = wait for a human.
  final int? autoReturnMs;

  /// Offered under the result.
  final List<String> actions;

  const Decision({
    required this.outcome,
    required this.tone,
    required this.admittedCount,
    required this.headline,
    this.detail,
    this.invitation,
    this.remaining,
    this.choices,
    required this.log,
    required this.autoReturnMs,
    required this.actions,
  });
}

const _admitDwell = 1500;
const _partialDwell = 2500;

Decision _deny(
  Outcome outcome,
  String headline,
  String? detail,
  List<String> actions,
) =>
    Decision(
      outcome: outcome,
      tone: Tone.deny,
      admittedCount: 0,
      headline: headline,
      detail: detail,
      log: true,
      autoReturnMs: null,
      actions: actions,
    );

Decision _hold(
  Outcome outcome,
  String headline,
  String? detail,
  List<String> actions, [
  LocalInvitation? invitation,
]) =>
    Decision(
      outcome: outcome,
      tone: Tone.hold,
      admittedCount: 0,
      headline: headline,
      detail: detail,
      invitation: invitation,
      log: true,
      autoReturnMs: null,
      actions: actions,
    );

Decision decide(Context ctx, ScanInput input) {
  LocalInvitation? inv;

  switch (input) {
    case ScanRaw(:final raw):
      // ---- 1. decode + signature, against every held key -----------------
      final v = verifyToken(raw, ctx.keys);

      if (v is VerifyFail) {
        if (v.reason == 'stale_version') {
          return _deny(
            Outcome.revoked,
            'Pass no longer valid',
            'This code was replaced. The guest needs their new link.',
            ['Search by name', 'Dismiss'],
          );
        }
        return _deny(
          Outcome.invalid,
          'Not a valid pass',
          "The code didn't verify. It may be a screenshot of something else, or damaged.",
          ['Search by name', 'Dismiss'],
        );
      }

      final ok = v as VerifyOk;

      // ---- 2. right event? ------------------------------------------------
      if (ok.payload.eventId != ctx.currentEventId) {
        return _deny(
          Outcome.wrongEvent,
          'Pass is for another event',
          'This pass belongs to ${ok.matched.eventName}.',
          ['Dismiss'],
        );
      }

      // ---- 3. on this leg's list? -----------------------------------------
      inv = ctx.find(ok.payload.passId);
      if (inv == null) {
        // Genuine pass for this event, but not invited to the leg being
        // scanned — the Abuja list at the Lagos gate.
        return _deny(
          Outcome.wrongLeg,
          'Not invited to this one',
          'This pass is for a different part of the event.',
          ['Search by name', 'Dismiss'],
        );
      }

    case ManualInput(:final passId):
      inv = ctx.find(passId);
      if (inv == null) {
        return _deny(
          Outcome.notFound,
          'No matching guest',
          'Nothing on the list for this event.',
          ['Add walk-in', 'Dismiss'],
        );
      }
  }

  // ---- 4. revoked --------------------------------------------------------
  if (inv.revoked) {
    return _deny(
      Outcome.revoked,
      'Pass revoked',
      '${inv.displayName} was removed from the guest list.',
      ['Call manager', 'Dismiss'],
    );
  }

  // ---- 5. RSVP gate ------------------------------------------------------
  if (inv.rsvp == Rsvp.declined) {
    return _hold(
      Outcome.rsvpDeclined,
      'They replied no',
      '${inv.displayName} declined this invitation.',
      ctx.canOverrideRsvp
          ? ['Admit anyway', 'Dismiss']
          : ['Call manager', 'Dismiss'],
      inv,
    );
  }
  if (ctx.policy.requireRsvp && inv.rsvp == Rsvp.pending) {
    return _hold(
      Outcome.rsvpBlocked,
      'No reply on file',
      'This event requires an RSVP before entry.',
      ctx.canOverrideRsvp
          ? ['Admit anyway', 'Dismiss']
          : ['Call manager', 'Dismiss'],
      inv,
    );
  }

  // ---- 6. allowance ------------------------------------------------------
  final remaining =
      (inv.allowance - inv.admitted) < 0 ? 0 : inv.allowance - inv.admitted;

  if (remaining == 0 && !ctx.policy.allowOverflow) {
    return _hold(
      Outcome.allowanceExhausted,
      'Party fully admitted',
      '${inv.admitted} of ${inv.allowance} already in.',
      ['Call manager', 'Dismiss'],
      inv,
    );
  }

  // ---- 7. how many arrived? ---------------------------------------------
  final asked = input.requestedCount;

  if (asked == null) {
    // Allowance of one needs no prompt. This is most guests, and this path
    // must never gain a tap.
    if (inv.allowance == 1 && remaining == 1) {
      return _applyAdmission(ctx, inv, 1, input, remaining);
    }
    if (remaining == 0) {
      // Overflow is allowed, so offer it rather than refusing outright.
      return Decision(
        outcome: Outcome.needsCount,
        tone: Tone.ask,
        admittedCount: 0,
        headline: inv.displayName,
        detail:
            '${inv.admitted} of ${inv.allowance} already admitted. Anyone else is over the invitation.',
        invitation: inv,
        remaining: 0,
        choices: const [1, 2, 3],
        log: false,
        autoReturnMs: null,
        actions: const [],
      );
    }
    final parts = [inv.category, inv.tableName].whereType<String>().join(' · ');
    return Decision(
      outcome: Outcome.needsCount,
      tone: Tone.ask,
      admittedCount: 0,
      headline: inv.displayName,
      detail: parts.isEmpty ? null : parts,
      invitation: inv,
      remaining: remaining,
      // Pre-selected at the full remaining count by the UI, so the common
      // case stays one tap.
      choices: List<int>.generate(remaining, (i) => i + 1),
      log: false,
      autoReturnMs: null,
      actions: const [],
    );
  }

  if (asked < 1) {
    return _deny(
      Outcome.invalid,
      'Nothing to admit',
      'Count must be at least one.',
      ['Dismiss'],
    );
  }

  return _applyAdmission(ctx, inv, asked, input, remaining);
}

Decision _applyAdmission(
  Context ctx,
  LocalInvitation inv,
  int count,
  ScanInput kind,
  int remaining,
) {
  final over = (count - remaining) < 0 ? 0 : count - remaining;

  if (over > 0 && !ctx.policy.allowOverflow) {
    return _deny(
      Outcome.overflowBlocked,
      'More people than invited',
      'Invited for ${inv.allowance}, ${inv.admitted + count} would be in. A manager must raise the invitation.',
      ['Call manager', 'Dismiss'],
    );
  }

  final afterRaw = inv.allowance - (inv.admitted + count);
  final after = afterRaw < 0 ? 0 : afterRaw;

  if (over > 0) {
    return Decision(
      outcome: Outcome.overflowAdmitted,
      tone: Tone.hold,
      admittedCount: count,
      headline: '$count admitted · $over over',
      detail:
          '${inv.displayName} was invited for ${inv.allowance}. The organiser has been notified.',
      invitation: inv,
      remaining: 0,
      log: true,
      autoReturnMs: _partialDwell,
      actions: const ['Undo'],
    );
  }

  final isManual = kind is ManualInput;
  final outcome = isManual
      ? Outcome.manual
      : count < remaining
          ? Outcome.partial
          : Outcome.admitted;

  final partial = after > 0;

  return Decision(
    outcome: outcome,
    tone: Tone.admit,
    admittedCount: count,
    headline: partial
        ? '$count of ${inv.allowance} admitted'
        : isManual
            ? 'Checked in by hand'
            : 'Admitted',
    detail: [inv.displayName, inv.tableName].whereType<String>().join(' · '),
    invitation: inv,
    remaining: after,
    log: true,
    // Longer when there is a number to read.
    autoReturnMs: partial ? _partialDwell : _admitDwell,
    actions: const ['Undo'],
  );
}

File: packages/checkin-core/src/checkin.ts
import { verifyToken, type EventKey } from "./token.ts";

/**
 * The check-in state machine.
 *
 * Deliberately pure: no database, no clock, no network. Everything it needs
 * arrives in the context. That is what lets the identical logic run on a
 * scanner phone with no signal and on the server when the queue syncs hours
 * later — and it is what makes the whole thing testable against the twelve
 * outcomes without standing up Postgres.
 *
 * Port of this file to Dart is the only logic duplicated across languages.
 * The test cases port with it.
 */

export type Rsvp = "pending" | "attending" | "partial" | "declined";

export type Outcome =
  | "needs_count"
  | "admitted"
  | "partial"
  | "overflow_admitted"
  | "manual"
  | "allowance_exhausted"
  | "invalid"
  | "wrong_event"
  | "wrong_leg"
  | "revoked"
  | "rsvp_blocked"
  | "rsvp_declined"
  | "overflow_blocked"
  | "event_cancelled"
  | "not_found";

export type Tone = "admit" | "hold" | "deny" | "ask";

/** One household's row in the scanner's local copy of the guest list. */
export type LocalInvitation = {
  passId: string;
  invitationId: string;
  eventId: string;
  legId: string;
  displayName: string;
  category: string | null;
  tableName: string | null;
  allowance: number;
  /** Derived locally: SUM(admitted_count) for this pass at this leg. */
  admitted: number;
  rsvp: Rsvp;
  revoked: boolean;
};

export type Policy = {
  allowOverflow: boolean;
  requireRsvp: boolean;
  /**
   * The organiser called the event off. Settings promises the guest a
   * cancellation notice AND that passes stop opening the gate, so this
   * refuses everything — the one refusal that outranks identifying who is
   * standing there, because the answer is the same for all of them.
   *
   * Not a billing block (HANDOFF §3 forbids those). Cancelling is a
   * deliberate act by the organiser, and reversible: set the event back to
   * active and every pass works again, because nothing was reissued.
   */
  eventCancelled: boolean;
};

export type Context = {
  currentEventId: string;
  currentLegId: string;
  policy: Policy;
  /** Keys for every event this usher works, not only the current one. */
  keys: EventKey[];
  /** Local lookup by pass id. Undefined means not in this leg's list. */
  find: (passId: string) => LocalInvitation | undefined;
  /** Usher permission. Off by default. */
  canOverrideRsvp: boolean;
};

export type ScanInput =
  | { kind: "scan"; raw: string; requestedCount?: number }
  /** Name search, then check in by hand. Skips all token checks. */
  | { kind: "manual"; passId: string; requestedCount?: number };

export type Decision = {
  outcome: Outcome;
  tone: Tone;
  /** Counts against allowance. Zero for every refusal. */
  admittedCount: number;
  headline: string;
  detail?: string;
  /** Household, when we could identify one. */
  invitation?: LocalInvitation;
  /** Remaining after this decision is applied. */
  remaining?: number;
  /** Present on needs_count — what the picker should offer. */
  choices?: number[];
  /** Whether to write a row. Refusals are logged; the count prompt is not. */
  log: boolean;
  /** Milliseconds before returning to the camera. null = wait for a human. */
  autoReturnMs: number | null;
  /** Offered under the result. */
  actions: string[];
};

const ADMIT_DWELL = 1500;
const PARTIAL_DWELL = 2500;

function deny(
  outcome: Outcome,
  headline: string,
  detail: string | undefined,
  actions: string[],
): Decision {
  return {
    outcome,
    tone: "deny",
    admittedCount: 0,
    headline,
    detail,
    log: true,
    autoReturnMs: null,
    actions,
  };
}

function hold(
  outcome: Outcome,
  headline: string,
  detail: string | undefined,
  actions: string[],
  invitation?: LocalInvitation,
): Decision {
  return {
    outcome,
    tone: "hold",
    admittedCount: 0,
    headline,
    detail,
    invitation,
    log: true,
    autoReturnMs: null,
    actions,
  };
}

export function decide(ctx: Context, input: ScanInput): Decision {
  let inv: LocalInvitation | undefined;

  // ---- 0. is there still an event? --------------------------------------
  // Ahead of the token checks on purpose: a cancelled event admits nobody,
  // so what they are holding does not matter, and an usher who has not
  // heard the news needs to be told the reason rather than "not a valid
  // pass". Manual check-in is refused too — otherwise Search by name is a
  // way around it.
  if (ctx.policy.eventCancelled) {
    return deny(
      "event_cancelled",
      "Event cancelled",
      "This event was called off. Nobody is being admitted — send them to the organiser.",
      ["Call manager", "Dismiss"],
    );
  }

  if (input.kind === "scan") {
    // ---- 1. decode + signature, against every held key -----------------
    const v = verifyToken(input.raw, ctx.keys);

    if (!v.ok) {
      if (v.reason === "stale_version") {
        return deny(
          "revoked",
          "Pass no longer valid",
          "This code was replaced. The guest needs their new link.",
          ["Search by name", "Dismiss"],
        );
      }
      return deny(
        "invalid",
        "Not a valid pass",
        "The code didn't verify. It may be a screenshot of something else, or damaged.",
        ["Search by name", "Dismiss"],
      );
    }

    // ---- 2. right event? ------------------------------------------------
    if (v.payload.eventId !== ctx.currentEventId) {
      return deny(
        "wrong_event",
        "Pass is for another event",
        `This pass belongs to ${v.matched.eventName}.`,
        ["Dismiss"],
      );
    }

    // ---- 3. on this leg's list? -----------------------------------------
    inv = ctx.find(v.payload.passId);
    if (!inv) {
      // Genuine pass for this event, but not invited to the leg being
      // scanned — the Abuja list at the Lagos gate.
      return deny(
        "wrong_leg",
        "Not invited to this one",
        "This pass is for a different part of the event.",
        ["Search by name", "Dismiss"],
      );
    }
  } else {
    inv = ctx.find(input.passId);
    if (!inv) {
      return deny("not_found", "No matching guest", "Nothing on the list for this event.", [
        "Add walk-in",
        "Dismiss",
      ]);
    }
  }

  // ---- 4. revoked --------------------------------------------------------
  if (inv.revoked) {
    return deny("revoked", "Pass revoked", `${inv.displayName} was removed from the guest list.`, [
      "Call manager",
      "Dismiss",
    ]);
  }

  // ---- 5. RSVP gate ------------------------------------------------------
  if (inv.rsvp === "declined") {
    return hold(
      "rsvp_declined",
      "They replied no",
      `${inv.displayName} declined this invitation.`,
      ctx.canOverrideRsvp ? ["Admit anyway", "Dismiss"] : ["Call manager", "Dismiss"],
      inv,
    );
  }
  if (ctx.policy.requireRsvp && inv.rsvp === "pending") {
    return hold(
      "rsvp_blocked",
      "No reply on file",
      "This event requires an RSVP before entry.",
      ctx.canOverrideRsvp ? ["Admit anyway", "Dismiss"] : ["Call manager", "Dismiss"],
      inv,
    );
  }

  // ---- 6. allowance ------------------------------------------------------
  const remaining = Math.max(0, inv.allowance - inv.admitted);

  if (remaining === 0 && !ctx.policy.allowOverflow) {
    return hold(
      "allowance_exhausted",
      "Party fully admitted",
      `${inv.admitted} of ${inv.allowance} already in.`,
      ["Call manager", "Dismiss"],
      inv,
    );
  }

  // ---- 7. how many arrived? ---------------------------------------------
  const asked = input.requestedCount;

  if (asked === undefined) {
    // Allowance of one needs no prompt. This is most guests, and this path
    // must never gain a tap.
    if (inv.allowance === 1 && remaining === 1) {
      return applyAdmission(ctx, inv, 1, input.kind, remaining);
    }
    if (remaining === 0) {
      // Overflow is allowed, so offer it rather than refusing outright.
      return {
        outcome: "needs_count",
        tone: "ask",
        admittedCount: 0,
        headline: inv.displayName,
        detail: `${inv.admitted} of ${inv.allowance} already admitted. Anyone else is over the invitation.`,
        invitation: inv,
        remaining: 0,
        choices: [1, 2, 3],
        log: false,
        autoReturnMs: null,
        actions: [],
      };
    }
    return {
      outcome: "needs_count",
      tone: "ask",
      admittedCount: 0,
      headline: inv.displayName,
      detail: [inv.category, inv.tableName].filter(Boolean).join(" · ") || undefined,
      invitation: inv,
      remaining,
      // Pre-selected at the full remaining count by the UI, so the common
      // case stays one tap.
      choices: Array.from({ length: remaining }, (_, i) => i + 1),
      log: false,
      autoReturnMs: null,
      actions: [],
    };
  }

  if (asked < 1) {
    return deny("invalid", "Nothing to admit", "Count must be at least one.", ["Dismiss"]);
  }

  return applyAdmission(ctx, inv, asked, input.kind, remaining);
}

function applyAdmission(
  ctx: Context,
  inv: LocalInvitation,
  count: number,
  kind: "scan" | "manual",
  remaining: number,
): Decision {
  const over = Math.max(0, count - remaining);

  if (over > 0 && !ctx.policy.allowOverflow) {
    return deny(
      "overflow_blocked",
      "More people than invited",
      `Invited for ${inv.allowance}, ${inv.admitted + count} would be in. A manager must raise the invitation.`,
      ["Call manager", "Dismiss"],
    );
  }

  const after = Math.max(0, inv.allowance - (inv.admitted + count));

  if (over > 0) {
    return {
      outcome: "overflow_admitted",
      tone: "hold",
      admittedCount: count,
      headline: `${count} admitted · ${over} over`,
      detail: `${inv.displayName} was invited for ${inv.allowance}. The organiser has been notified.`,
      invitation: inv,
      remaining: 0,
      log: true,
      autoReturnMs: PARTIAL_DWELL,
      actions: ["Undo"],
    };
  }

  const outcome: Outcome =
    kind === "manual" ? "manual" : count < remaining ? "partial" : "admitted";

  const partial = after > 0;

  return {
    outcome,
    tone: "admit",
    admittedCount: count,
    headline: partial
      ? `${count} of ${inv.allowance} admitted`
      : kind === "manual"
        ? "Checked in by hand"
        : "Admitted",
    detail: [inv.displayName, inv.tableName].filter(Boolean).join(" · "),
    invitation: inv,
    remaining: after,
    log: true,
    // Longer when there is a number to read.
    autoReturnMs: partial ? PARTIAL_DWELL : ADMIT_DWELL,
    actions: ["Undo"],
  };
}
